presto-main/src/main/java/com/facebook/presto/operator/SpatialJoinOperator.java

```diff
@@ -0,0 +1,244 @@
+        {
+            return new SpatialJoinOperatorFactory(operatorId, planNodeId, probeTypes, probeOutputChannels, probeGeometryChannel, pagesSpatialIndexFactory, spatialRelationshipTest);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final List<Type> probeTypes;
+    private final List<Integer> probeOutputChannels;
+    private final int probeGeometryChannel;
+    private final ListenableFuture<PagesSpatialIndex> pagesSpatialIndexFuture;
+    private final BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest;
+    private final List<Type> outputTypes;
+    private final PageBuilder pageBuilder;
+    private Page probe;
+    private int probePosition;
+    private List<Long> joinAddresses;
+    private int joinAddressIndex;
+    private boolean finishing;
+    private boolean finished;
+
+    public SpatialJoinOperator(
+            OperatorContext operatorContext,
+            List<Type> outputTypes,
+            List<Type> probeTypes,
+            List<Integer> probeOutputChannels,
+            int probeGeometryChannel,
+            ListenableFuture<PagesSpatialIndex> pagesSpatialIndexFuture,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest)
+    {
+        this.operatorContext = operatorContext;
+        this.probeTypes = ImmutableList.copyOf(probeTypes);
+        this.probeOutputChannels = ImmutableList.copyOf(probeOutputChannels);
+        this.probeGeometryChannel = probeGeometryChannel;
+        this.pagesSpatialIndexFuture = pagesSpatialIndexFuture;
+        this.outputTypes = ImmutableList.copyOf(outputTypes);
+        this.pageBuilder = new PageBuilder(outputTypes);
+        this.spatialRelationshipTest = spatialRelationshipTest;
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return outputTypes;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return pagesSpatialIndexFuture.isDone() && !pageBuilder.isFull() && probe == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        verify(probe == null);
+        probe = page;
+        probePosition = 0;
+
+        joinAddresses = null;
+        joinAddressIndex = 0;
+    }
+
+    private void processProbe()
+    {
+        verify(probe != null);
+
+        PagesSpatialIndex pagesSpatialIndex = getDone(pagesSpatialIndexFuture);
+        for (int position = probePosition; position < probe.getPositionCount(); position++) {
+            if (this.joinAddresses == null) {
+                joinAddresses = pagesSpatialIndex.findJoinAddresses(position, probe, probeGeometryChannel, spatialRelationshipTest);
+                joinAddressIndex = -1;
+            }
+
+            for (int joinAddressIndex = this.joinAddressIndex + 1; joinAddressIndex < joinAddresses.size(); joinAddressIndex++) {
+                if (pageBuilder.isFull()) {
```

Comment:
You're existing this only when page builder full. However, when many rows don't actually match, you can spend a lot of time here without returning. Operators shouldn't do this and for sake of that you should consult return also when `com.facebook.presto.operator.DriverYieldSignal#isSet`.